uplink/storage/streams/store.go

```diff
@@ -137,18 +135,24 @@ func (s *streamStore) upload(ctx context.Context, path Path, pathCipher storj.Ci
 	}
 
 	var (
-		streamID       storj.StreamID
-		currentSegment int64
+		streamID         storj.StreamID
+		uploadedSegments int64
 	)
 	defer func() {
+		if err != nil {
+			s.cancelHandler(context.Background(), streamID, uploadedSegments, path, pathCipher)
+			return
+		}
+
 		select {
 		case <-ctx.Done():
-			s.cancelHandler(context.Background(), streamID, currentSegment, path, pathCipher)
+			s.cancelHandler(context.Background(), streamID, uploadedSegments, path, pathCipher)
 		default:
 		}
 	}()
 
 	var (
+		currentSegment       int64 = -1
```

Comment:
Initializing a counter to `-1` is not common. What was wrong with initializing it with `0` and incrementing its value at the end of the for-cycle?

Reply:
That was the problem is that if it gets incremented at the end of the loop in the last iteration is when it gets incremented but there are no more segments because of `!eofReader.isEOF() && !eofReader.hasError()`.

Then the variable is called `currentSegment` but when the loop finishes, it isn't the current segment, it's one more.

I can start with 0 and just after the loop subtracts one but anyway isn't clear either.

Before I introduced the `uploadedSegments` variable, starting with `-1` made more sense because `currentSegment` was passed to the `cancelHandler` and depending where the error happened the value was incremented by 1 when it shouldn't be, but with `uploadedSegmetns` it doesn't happen anymore.

I can add a comment or starts with 0, do the subtract at the end but anyway that will require a comment to explain why, too.

Reply:
Another solution, which I've just thought is at the end of the loop, to declare a new variable called `numberOfSegments` and assign the `currentSegment` and use it for the logic executed at the end of the function outside of the loop.